### PR TITLE
feat: Added Round Robin CPU scheduling algorithm

### PR DESCRIPTION
**Pull request by varadpundlik**
### Round Robin Algorithm
Added implementation of Round Robin CPU scheduling algorithm in directory ./cpu_shceduling_algorithms

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Round Robin is a CPU scheduling algorithm where each process is assigned a fixed time slot in a cyclic way. It is basically the preemptive version of First come First Serve CPU Scheduling algorithm.

### CODE REVIEW

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ *
+ */
+
+#include <algorithm>  // for max function
+#include <cassert>    // fot testing purpose
+#include <iostream>   //for printing results
+#include <queue>      //for job queue
+#include <string>     // for job name
+#include <utility>    //for using pair
+#include <vector>     // for tasklist of processes
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:19]:
```suggestion
#include <algorithm>  /// for max function
#include <cassert>    /// for testing purpose
#include <iostream>   /// for printing results
#include <queue>      /// for job queue
#include <string>     /// for job name
#include <utility>    /// for using pair
#include <vector>     /// for tasklist of processes
```

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+*/
+class Job {
+    int id;
+    string name;
+    int at;   // Arrival time
+    int wt{0};  // Waiting time
+    int bt;   // Burst time
+    int b1;   // Remaining burst time
+    int tat{0};  // Turnaround time
+    int ct{0};   // Completion time
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:41]:
```suggestion
    int at;   ///< Arrival time
    int wt{0};  ///< Waiting time
    int bt;   ///< Burst time
    int b1;   ///< Remaining burst time
    int tat{0};  ///< Turnaround time
    int ct{0};   ///< Completion time
```

Please make these variable identifiers longer and detailed. currently you would need to glance at documentation just to understand what the variables stand for

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    }
+};
+/** @brief functoin for testing round robin algorithm
+ *  @param input number of jobs and each job's id name arrival time and burst
+ * time
+ * @returns void
+ */
+
+void test() {
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:166]:
```suggestion
static void test() {
```

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                 << "\t" << i.ct << "\t" << i.wt << "\t" << i.tat << endl;
+        }
+    }
+};
+/** @brief functoin for testing round robin algorithm
+ *  @param input number of jobs and each job's id name arrival time and burst
+ * time
+ * @returns void
+ */
+
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:165]:
```suggestion
```

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    pair<double, double> exp = {4.75, 7.75};
+    assert(p == exp);
+    cout << "Test case successfully passed!!!" << endl;
+}
+
+/** @brief entrypoint of program
+ *  @returns 0
+ */
+int main() {
+    test();
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:189]:
```suggestion
    test(); // run self-test implimentations
```

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+/** @class Schedduler
+ *  @brief class to performs schedulig algorithms
+ *  @param taskList List of processes
+ *  @param num number of jobs
+ */
+class Scheduler {
+    int num;               // Number of jobs
+    vector<Job> taskList;  // List of jobs
+    double avgWT{0};       // Average waiting time
+    double avgTAT{0};      // Average turnaround time
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:69]:
```suggestion
    int num;               ///< Number of jobs
    vector<Job> taskList;  ///< List of jobs
    double avgWT{0};       ///< Average waiting time
    double avgTAT{0};      ///< Average turnaround time
```

as mentioned above, variable names with more meaning

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you for your contributions

File: cpu_scheduling_algorithms/round_robin.cpp (hunk 6 above)
Comment (varadpundlik) [reply cpu_scheduling_algorithms/round_robin.cpp]: updated variable names in new commit

File: cpu_scheduling_algorithms/round_robin.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+ * time
+ * @returns void
+ */
+static void test() {
+    vector<Job> t;
+    Job j1(1, "P1", 0, 5);
+    Job j2(2, "P2", 1, 4);
+    Job j3(3, "P3", 2, 2);
+    Job j4(4, "P4", 4, 1);
+    t.push_back(j1);
```
Comment (realstealthninja) [inline cpu_scheduling_algorithms/round_robin.cpp:188]: add test cases for negative numbers or use unsigned in your algorithm

Comment (varadpundlik) [reply cpu_scheduling_algorithms/round_robin.cpp:187]: Converted time related variables to unsigned int datatype in new commit

**Review by realstealthninja — APPROVED** [review]
This looks amazing thank you for your contributions.

**Pull request closed without merging**